### PR TITLE
Module object refactoring

### PR DESCRIPTION
**Pull request by soc**

### CODE REVIEW

**Comment by soc** [comment]
See discussion: https://groups.google.com/d/topic/scala-internals/vEFtAdiybyk/discussion

**Comment by xeno-by** [comment]
1) There are still traces of "module", e.g. in Flags, Modes and some tests (the list is non-exhaustive).
2) `objct` sounds weird. I'd go for `obj`
3) Why again can't we rename MODULE$ to INSTANCE$?

All in all, the surface of changes is surprisingly small. If you re-introduce and deprecate old module-based names, I'm more or less positive about this change.

**Comment by soc** [comment]
@xeno-by:

1) Some things might be on purpose and some might be forgotten. I'll have a look.
2) Absolutely right. I wonder what I was thinking when I came up with `objct`. :-)
3) Someone (was it @paulp? ... not sure) suggested not to modify code which changes stuff written to class files.

The deprecated-modules commit is being tested right now. :-)

**Comment by soc** [comment]
I see a few “Modules” in the MSIL code which I didn't touch because I'm not that familiar with the CLR terminology. Maybe @magarciaEPFL can suggest the changes he would like to see (if there are any)?

**Comment by soc** [comment]
PLS REBUILD ALL

File: src/compiler/scala/tools/nsc/matching/Patterns.scala [hunk 1]
```diff
@@ -456,9 +456,9 @@ trait Patterns extends ast.TreeDSL {
     def    tpe  = tree.tpe
     def isEmpty = tree.isEmpty
 
-    def isModule    = sym.isModule || tpe.termSymbol.isModule
+    def isObject    = sym.isObject || tpe.termSymbol.isObject
     def isCaseClass = tpe.typeSymbol.isCase
-    def isObject    = (sym != null) && (sym != NoSymbol) && tpe.prefix.isStable  // XXX not entire logic
+    def isInstance  = (sym != null) && (sym != NoSymbol) && tpe.prefix.isStable  // XXX not entire logic
```
Comment (soc) [inline src/compiler/scala/tools/nsc/matching/Patterns.scala:461]: Is this renaming OK or is there a better name?

**Comment by soc** [comment]
PLS REBUILD ALL

**Comment by adriaanm** [comment]
on a conceptual level, this is a discussion worth having, but unfortunately it's much too late in the 2.10.0 release cycle
if we decide to rename module (most likely to singleton since object is too vague -- what's an ObjectClass?),
we can deprecate object in the 2.10.x cycle and have the fundamental discussion on its new name then

**Comment by soc** [comment]
Well, I think that would combine all the drawbacks of keeping the name with the all drawbacks of changing the name. I don't think that's worth it. (The reflection API was supposed to be stable, right?)

**Comment by adriaanm** [comment]
"stable" doesn't mean "set in stone", it means "deprecation required before changes"

**Comment by adriaanm** [comment]
oh, it also has all the advantages of not changing the name (as in avoiding last-minute regressions in e.g., the IDE/SBT and other tools that are sensitive to this API)

this is not a critical bug. it's not a regression. it's too late. sorry

**Comment by soc** [comment]
Sure, no problem.

**Pull request closed without merging**